### PR TITLE
fix(iot-deps): Fixed the issue by modifying MethodParser class

### PR DESCRIPTION
**Pull request by davilu**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-java/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [x] This pull-request is submitted against the `master` branch. 

# Reference/Link to the issue solved with this PR (if any)
https://github.com/Azure/azure-iot-sdk-java/issues/378
# Description of the problem

[Device Client] DeviceMethodCallback message contains backslashes
# Description of the solution

Fixed the issue by modifying MethodParser class

### CODE REVIEW

**Comment by jasmineymlo** [comment]
Would you fill me in with some details on your fix in MethodParser class?  like why are we replacing Gson, is the output from it not what we are expecting?  Thanks.

**Comment by davilu** [comment]
`Gson` is a tool class used for serialize/deserialize known class type. In this case, payload is send/received as String. `Gson.toJsonTree`  will just json escape `payload`. We'd like to parse `payload` to a `JsonElement` and send as `payload` attribute which is `JsonParser`'s functionality.  For example, if Service client invoke `methodName` as `test` and payload as `{"a":"A"}`, `payload` will be escaped by `Gson` as `"{\"a\": \"A\"}` and be sent as:
```
{
     "methodName": "test",
     "payload": "{\"a\": \"A\"}"
}
```
`JsonParser` will convert it to a `JsonObject`, serialize as `{"abc":"XYZ"}` and sent as
```
{
	"methodName": "test",
	"payload": {
		"a": "A"
	}
}
```
IoT Hub does a good job and just convert `payload` as `JsonElement` to `String`, which will be `"{\"a\": \"A\"}"` with former `Gson` code and `{"a":"A"}` with the fix. 

Once device client receives the `payload`, it will pass the `payload` as `String` to Application `DeviceMethodCallback` API. 

**Caution** Application will get `payload` as it is sent out now if it's a valid json string. Ideally `payload` is required to be json string but for back compatible, any invalid string will be accepted, json escaped and received. For example if `payload` is `abc` will is invalid, it will be received as `"abc"` on `DeviceMethodCallback` API. 
For Java specific, `Map<String,Object>` will be converted as `JsonObject`, e.g. following map
```
{"boolean"=true, "string"="STRING", "int"=123, "map"= { "internal": 1.3}}
```
will be sent and received as:
```
{
     "boolean": true,
     "string": "STRING",
     "int": 123
     "map": {
          "internal": 1.3
     }
}
```
  Basic class type `Number` and `Boolean will be received as it's value. Any other `Object` will be called `toString()` and follow the process above. 

After `DeviceMethodCallback` process, when `DeviceMethodCallback` send back `payload` as `String`, it will reverse the process above, Json `String` will be converted to `Map`.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java [hunk 1]
```diff
@@ -286,6 +320,48 @@ public String toJson() throws IllegalArgumentException
...
+                }
+                else
+                {
+                    return jsonElement;
+                }
+            }
+            catch (JsonSyntaxException e)
+            {
+                return new Gson().toJsonTree(payload);
+            }
```
Comment (prmathur-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: I like this change!

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java [hunk 2]
```diff
@@ -3,11 +3,23 @@
```
Comment (prmathur-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: Commit description 
**fix(iot-deps):** Fixed the issue by modifying MethodParser class 

I don't see any changes in device

Comment (davilu) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: No change required on device client side.

**Comment by maxtar** [comment]
Hi, when we can expect fix?

**Comment by davilu** [comment]
>           Hi, when we will expect fix?

@maxtar It's been finalized and will be merged soon.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java (hunk 2 above)
Comment (prmathur-microsoft) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: then please change the commit message to reflect deps

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java [hunk 3]
```diff
@@ -148,7 +148,7 @@ private synchronized MethodResult invokeMethod(URL url, String methodName, Long
         }
 
         long  responseTimeout, connectTimeout;
-        
+
         if (responseTimeoutInSeconds == null)
```
Comment (prmathur-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java]: I want to make sure is there any breaking behavior change post this PR ? Also can you confirm Jobs and methods E2E pass with your change?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java (hunk 2 above)
Comment (davilu) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: done

**Comment by davilu** [comment]
The comments are all addressed.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java [hunk 4]
```diff
@@ -142,106 +154,124 @@ public MethodParser(Object payload)
...
+                            status = statusTagNode.getAsInt();
+                        }
+                        JsonElement payloadNode = jsonObject.get(PAYLOAD_TAG);
+                        if (payloadNode != null)
+                        {
+                            payload = resolveJsonElement(payloadNode);
+                        }
+                    }
+                }
+                else {
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: >{ [](start = 21, length = 1)

nit: Fix bracing in these if/else blocks. We currently do braces on next line

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java [hunk 5]
```diff
@@ -142,106 +154,124 @@ public MethodParser(Object payload)
...
+                    }
+                    else {
+                        /* Codes_SRS_METHODPARSER_21_008: [If the provided json is null, empty, or not valid, the fromJson shall throws IllegalArgumentException.] */
+                        throw new IllegalArgumentException("Invoke method name and Status reported in the same json");
+                    }
+                }
             }
+            else {
+                // JSON null
+                throw new IllegalArgumentException("Invalid json.");
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: >Invalid json [](start = 52, length = 12)

could be more descriptive. Why is the json invalid?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java [hunk 6]
```diff
@@ -142,106 +154,124 @@ public MethodParser(Object payload)
...
+            }
+        }
+        catch (Exception ex)
+        {
+            /* Codes_SRS_METHODPARSER_21_008: [If the provided json is null, empty, or not valid, the fromJson shall throws IllegalArgumentException.] */
+            throw new IllegalArgumentException("Malformed json:" + ex);
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: >ex [](start = 67, length = 2)

stacktrace is lost here

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java [hunk 7]
```diff
@@ -286,6 +320,48 @@ public String toJson() throws IllegalArgumentException
         return toJsonElement().toString();
     }
 
+    @SuppressWarnings("unchecked")
+    private JsonElement jsonizePayload(Object payload)
+    {
+        if (payload == null) {
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/serializer/MethodParser.java]: >{ [](start = 29, length = 1)

nit: bracing style again

**Comment by timtay-microsoft** [comment]
Just a few small comments. Besides them, I approve this PR

**Pull request closed without merging**